ouroboros-network-framework/src/Ouroboros/Network/Subscription/Dns.hs

```diff
@@ -87,11 +87,23 @@ dnsResolve tracer resolver peerStatesVar beforeConnect (DnsSubscriptionTarget do
     ipv4Rsps <- newEmptyTMVarM
     gotIpv6Rsp <- Lazy.newTVarM False
 
-    aid_ipv6 <- async $ resolveAAAA gotIpv6Rsp ipv6Rsps
-    aid_ipv4 <- async $ resolveA gotIpv6Rsp ipv4Rsps
+    res <-
+      race
+        (threadDelay 20)
+        (do
+           aid_ipv6 <- async $ resolveAAAA gotIpv6Rsp ipv6Rsps
+           aid_ipv4 <- async $ resolveA gotIpv6Rsp ipv4Rsps
+
+           rd_e <- waitEitherCatch aid_ipv6 aid_ipv4
+           handleResult ipv6Rsps ipv4Rsps rd_e
+        )
+    case res of
+      Left _ -> do
+        -- TODO: the thread timedout, we should trace it
+        return (SubscriptionTarget $ pure Nothing)
+      Right st ->
+        return st
```

Comment:
This would be simpler using `timeout` (and use fewer threads in the happy case).
```
  timeout :: DiffTime -> m a -> m (Maybe a)
```

```suggestion
    -- Though the DNS lib does have its own timeouts, these do not work
    -- on Windows reliably so as a workaround we add an extra layer
    -- of timeout on the outside.
    -- TODO: fix upstream dns lib
    res <- timeout 20 $ do
             aid_ipv6 <- async $ resolveAAAA gotIpv6Rsp ipv6Rsps
             aid_ipv4 <- async $ resolveA gotIpv6Rsp ipv4Rsps
             rd_e <- waitEitherCatch aid_ipv6 aid_ipv4
             handleResult ipv6Rsps ipv4Rsps rd_e
    case res of
      Nothing -> do
        -- TODO: the thread timedout, we should trace it
        return (SubscriptionTarget $ pure Nothing)
      Just st ->
        return st
```

Also, this is suspicious from the existing code:
```
             aid_ipv4 <- async $ resolveA gotIpv6Rsp ipv4Rsps
```
Shouldn't that use `gotIpv4Rsp` not 6?

Reply:
That was my __todo_ :)

Reply:
No, it should be `gotIpv6Rsp`. `gotIpv6Rsp` is used for implementing the short delay used to give preferential treatment to IPv6.
` {- From RFC8305.
     - If a positive A response is received first due to reordering, the client
      - SHOULD wait a short time for the AAAA response to ensure that preference is
      - given to IPv6.
 -}`

Reply:
:+1: OK.